tests/integration/frameworks/test_transformers_unit.py

```diff
@@ -0,0 +1,217 @@
+
+
+def test_raise_does_not_match_task_name():
+    # pipeline task does not match given task name or pipeline.task is None
+    with pytest.raises(
+        BentoMLException,
+        match=f"Argument `task_name` 'custom' does not match pipeline task name '{sentiment.task}'.",
+    ):
+        _ = bentoml.transformers.save_model(
+            "forbidden_override",
+            sentiment,
+            task_name="custom",
+            task_definition=original_task,  # type: ignore (unfinished transformers type)
+        )
+
+
+def test_raise_does_not_match_impl_field():
+    # task_definition['impl'] is different from pipeline type
+    orig_impl: type = original_task["impl"]
+    try:
+        with pytest.raises(
+            BentoMLException,
+            match=f"Argument `pipeline` is not an instance of {AudioClassificationPipeline}. It is an instance of {type(sentiment)}.",
+        ):
+            original_task["impl"] = AudioClassificationPipeline
+            _ = bentoml.transformers.save_model(
+                "forbidden_override",
+                sentiment,
+                task_name=alias,
+                task_definition=original_task,  # type: ignore (unfinished transformers type)
+            )
+    finally:
+        original_task["impl"] = orig_impl  # type: ignore (unfinished transformers type)
+
+
+def test_raises_is_not_pipeline_instance():
+    with pytest.raises(BentoMLException) as exc_info:
+        _ = bentoml.transformers.save_model(
+            "not_pipeline_type", AudioClassificationPipeline  # type: ignore (testing invalid type)
+        )
+    assert (
+        "`pipeline` must be an instance of `transformers.pipelines.base.Pipeline`. "
+        in str(exc_info.value)
+    )
+
+
+def test_logs_custom_task_definition(caplog: pytest.LogCaptureFixture):
+    with caplog.at_level(logging.INFO):
+        _ = bentoml.transformers.save_model(
+            "custom_sentiment_pipeline",
+            sentiment,
+            task_name="sentiment-analysis",
+            task_definition=original_task,  # type: ignore (unfinished transformers type)
+        )
+    assert (
+        "Arguments `task_name` and `task_definition` are provided. Saving model with pipeline "
+        in caplog.text
+    )
+
+
+def test_model_options():
+    unstructured_options: t.Dict[str, t.Any] = {
+        "task": "sentiment-analysis",
+        "tf": [],
+        "pt": [],
+        "default": {},
+        "type": None,
+        "kwargs": {},
+    }
+
+    structured_options = ModelOptions(task="sentiment-analysis")
+    assert structured_options.to_dict() == unstructured_options
+
+    structured_options = ModelOptions(**unstructured_options)
+    assert structured_options is not None
+    assert structured_options.tf == ()
+    assert structured_options.pt == ()
+    assert structured_options.default == {}
+    assert structured_options.type is None
+    assert structured_options.kwargs == {}
+
+
+def test_custom_pipeline():
```

Comment:
For creating custom tasks and injecting it, we can  use a context manager, or a `try...except` loop to handle the lifecycle.